### PR TITLE
ethash: Optimize difficulty to boundary implementation

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: libethashseal/Ethash.h [hunk 1]
```diff
@@ -69,7 +69,13 @@ class Ethash: public SealEngineBase
     static h256 seedHash(BlockHeader const& _bi);
     static Nonce nonce(BlockHeader const& _bi) { return _bi.seal<Nonce>(NonceField); }
     static h256 mixHash(BlockHeader const& _bi) { return _bi.seal<h256>(MixHashField); }
-    static h256 boundary(BlockHeader const& _bi) { auto d = _bi.difficulty(); return d ? (h256)u256((bigint(1) << 256) / d) : h256(); }
+
+    static h256 boundary(BlockHeader const& _bi)
+    {
+        const auto& d = _bi.difficulty();
```
Comment (gumb0) [inline libethashseal/Ethash.h]: `auto const&`

File: libethashseal/Ethash.h [hunk 2]
```diff
@@ -69,7 +69,13 @@ class Ethash: public SealEngineBase
...
+        return d > 1 ? h256{u256((u512(1) << 256) / d)} : ~h256{};
```
Comment (gumb0) [inline libethashseal/Ethash.h]: It returned 0 before for d == 0, now it returns `ff...ff`, is this correct?

Comment (gumb0) [reply libethashseal/Ethash.h]: Maybe make a `constexpr` constant out of `~h256{}`

Comment (chfast) [reply libethashseal/Ethash.h]: > Maybe make a `constexpr` constant out of `~h256{}`

How? We don't have any `constexpr` support in `h256`.

Comment (gumb0) [reply libethashseal/Ethash.h]: Could be just const then.

Comment (chfast) [reply libethashseal/Ethash.h]: You mean `const h256 maxBoundary = ~h256{}`?

Comment (gumb0) [reply libethashseal/Ethash.h]: Yes. Then it's guaranteed not to calculate negation each time.

**Pull request merged**